### PR TITLE
Implement `HashToScalar` for P-384

### PR DESCRIPTION
**Pull request by daxpedda**
Depends on https://github.com/RustCrypto/traits/pull/1024.

### CODE REVIEW

File: p384/src/arithmetic/hash2curve.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            dst: b"HashToScalar-VOPRF08-\x00\x00\x04",
+            seed: &hex!("a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3"),
+            sk_sm: &hex!("b9ff42e68ef6f8eaa3b4d15d15ceb6f3f36b9dc332a3473d64840fc7b44626c6e70336bdecbe01d9c512b7e7d7e6af21"),
+        },
+        TestVector {
+            dst: b"HashToScalar-VOPRF08-\x01\x00\x04",
+            seed: &hex!("a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3a3"),
+            sk_sm: &hex!("42c4d1c15d27be015844404088967afe48c8ae96d4f00ce48e4d38ecabfb8feb5b748de625cdf81ab076745d6211be95"),
+        },
+    ];
```
Comment (daxpedda) [inline p384/src/arithmetic/hash2curve.rs:91]: I'm using test Voprf version 8 test vectors to keep it simple, as version 9 requires more code.

File: p384/src/arithmetic/hash2curve.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        let mut out = [0; 72];
+
+        for (src, dst) in scalar
+            .into_limbs()
+            .into_iter()
+            .rev()
+            .zip(out.chunks_exact_mut(Limb::BYTE_SIZE))
+        {
+            dst.copy_from_slice(&src.to_be_bytes());
+        }
```
Comment (daxpedda) [inline p384/src/arithmetic/hash2curve.rs:125]: This is just `ArrayEncoding::to_be_byte_array`, which isn't available for `U576`.

File: p384/src/arithmetic/scalar.rs [hunk 3]
```diff
@@ -35,7 +35,7 @@ impl ScalarArithmetic for NistP384 {
 /// Scalars are elements in the finite field modulo n.
 #[derive(Clone, Copy, Debug)]
 #[cfg_attr(docsrs, doc(cfg(feature = "arithmetic")))]
-pub struct Scalar(U384);
+pub struct Scalar(pub(crate) U384);
```
Comment (daxpedda) [inline p384/src/arithmetic/scalar.rs:38]: I hope this is okay.

File: p384/src/arithmetic/hash2curve.rs (hunk 2 above)
Comment (tarcieri) [reply p384/src/arithmetic/hash2curve.rs:125]: I can potentially add `U576`

File: p384/src/arithmetic/scalar.rs (hunk 3 above)
Comment (tarcieri) [reply p384/src/arithmetic/scalar.rs:38]: That's fine, although potentially avoidable if all of the constants are given in Montgomery form

File: p384/src/arithmetic/hash2curve.rs (hunk 2 above)
Comment (tarcieri) [reply p384/src/arithmetic/hash2curve.rs:125]: As it turns out, `U576` is useful for `p521` as well, so I went ahead and added it:

https://github.com/RustCrypto/crypto-bigint/pull/94

Comment (tarcieri) [reply p384/src/arithmetic/hash2curve.rs:125]: Released in `crypto-bigint` v0.4.5

File: p384/src/arithmetic/hash2curve.rs (hunk 1 above)
Comment (daxpedda) [reply p384/src/arithmetic/hash2curve.rs]: Now that everything's working, I updated the test vectors.

File: p384/src/arithmetic/hash2curve.rs (hunk 2 above)
Comment (daxpedda) [reply p384/src/arithmetic/hash2curve.rs]: Amazing, that helped out a lot. How do I pull in 0.4.5 though?
- Update `elliptic-curve` to require `0.4.5`.
- Add a direct dependency to `0.4.5`.
- Only update `Cargo.lock`. That wouldn't work for downstream users, right?

File: p384/src/arithmetic/scalar.rs (hunk 3 above)
Comment (daxpedda) [reply p384/src/arithmetic/scalar.rs]: Didn't need it anymore, so I reverted it.

**Comment by daxpedda** [comment]
So this has to build on https://github.com/RustCrypto/traits/pull/1024 now. Considering this will be a breaking change, what's the solution here?
I could just copy-paste the OSSWU code here, like we did in [`k256`](https://github.com/RustCrypto/elliptic-curves/blob/0dd58a00593cc77952ab7101cce284c4751a78f6/k256/src/arithmetic/hash2curve.rs#L85-L127). Or should we just delay this PR? Or release an `elliptic-curve` 0.13-pre?

**Comment by tarcieri** [comment]
I think it'd probably make sense to just delay the PR until the next round of breaking changes to `elliptic-curve`, and in the meantime perhaps we could work out some better solutions for https://github.com/RustCrypto/traits/pull/1024 regarding the relationships between `FieldElement`s and their inner `UInt`s.

**Comment by daxpedda** [comment]
Alright, good to know!
I'm in no rush :).

**Comment by daxpedda** [comment]
Updated to https://github.com/RustCrypto/traits/commit/28ec035531ed028ba010a652980f78007c8f49ff.

**Comment by daxpedda** [comment]
@tarcieri this is ready now.

**Review by tarcieri — APPROVED** [review]

**Pull request merged**